### PR TITLE
Support FastSpeech2 (+FastPitch)

### PR DESCRIPTION
**Pull request by kan-bayashi**
This PR adds FastSpeech2 module (#2019).

Changes:
- Add FastSpeech2 model.
- Add new arguments for TTS task.
	- `pitch_extractor`
	- `energy_extractor`
	- `pitch_normalize`
	- `energy_normalize`
- Add `Dio` for pitch extraction.
- Add `Energy` for energy exraction.
- Add `--write_collected_feats` option in `tts.sh` to speed up the training.
- Add `--tts_stats_dir` option in `tts.sh` to specify the stats dir from command line.

Note:
- Instead of quantized embedding of f0/energy, use continuous value.
- Use FastPitch-style token-averaged f0 / energy as discussed in #2019 

Todo:
- [x] Implement FastSpeech2
- [x] Add test for FastSpeech2
- [x] Implement F0 / energy extraction
- [x] Add test for f0 / energy extraction
- [x] Support in recipe level
- [x] check performance

Procedure to train fastspeech2
```bash
# 1. Train teacher model
./run.sh --train_config conf/tuning/train_tacotron2.yaml

# 2. Dump durations using teacher model with teacher forcing
./run.sh --stage 7 --train_config conf/tuning/train_tacotron2.yaml --test_sets "tr_no_dev dev eval1" --inference_args "--use_teacher_forcing true"

# 3. Calculate statistics pitch and energy features
./run.sh --stage 5 --stop-stage 5 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2
# To speed up the training, it is better to dump the features
./run.sh --stage 5 --stop-stage 5 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2 --write_collected_feats true

# 4. Train fastspeech2
./run.sh --stage 6 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2
```

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Almost done.

**Comment by kan-bayashi** [comment]
@kamo-naoyuki Could you check, especially the following points?
- New options in `tts.sh`
- New feature extractor and normalize in `espnet_model.py` and `tts.py`

@r9y9 Could you check `fastspeech2.py` and `dio.py`?

**Comment by r9y9** [comment]
Sure, I will check them tomorrow.

**Comment by kan-bayashi** [comment]
I uploaded some samples.
https://drive.google.com/drive/folders/18dZz5jP05Vh5bRRK8nT7HnmLXSeLsitm?usp=sharing

v1: baseline
v2: deep pitch predictor
v4: deep pitch predictor + stop gradient
gst: GST + v4

Compared to the FastSpeech samples, the quality is clearly improved.

File: espnet2/tts/feats_extract/dio.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+from scipy.interpolate import interp1d
+from typeguard import check_argument_types
+
+from espnet.nets.pytorch_backend.nets_utils import pad_list
+from espnet2.tts.feats_extract.abs_feats_extract import AbsFeatsExtract
+
+
+class Dio(AbsFeatsExtract):
+    """F0 estimation with dio + stonemask algortihm."""
```
Comment (sw005320) [inline espnet2/tts/feats_extract/dio.py:26]: - Maybe you can add more reference information about this (sorry if you put it somewhere)?
- It's better to state that this part cannot be a part of the computational graph and add TODO somewhere about the possibility of rewriting the f0 estimation like that.
- Do you think we can use it for ASR? If so, we'll test it.

File: espnet2/tts/feats_extract/energy.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
+# Copyright 2020 Nagoya University (Tomoki Hayashi)
```
Comment (sw005320) [inline espnet2/tts/feats_extract/energy.py:1]: This function might be duplicated in ASR and SE, but it has some unique methods (duration and frame length adjustment), and we can leave it as it is.

File: egs2/csmsc/tts1/conf/tuning/decode_tacotron2.yaml [hunk 3]
```diff
@@ -7,9 +7,10 @@
 ##########################################################
```
Comment (sw005320) [inline egs2/csmsc/tts1/conf/tuning/decode_tacotron2.yaml:7]: Should ASR/SE follow a similar style?
I like this style.

File: espnet2/tts/feats_extract/dio.py (hunk 1 above)
Comment (kan-bayashi) [reply espnet2/tts/feats_extract/dio.py:26]: > Maybe you can add more reference information about this (sorry if you put it somewhere)?

Sure!

> It's better to state that this part cannot be a part of the computational graph and add TODO somewhere about the possibility of rewriting the f0 estimation like that.

Agree. I will add a description about it.

> Do you think we can use it for ASR? If so, we'll test it.

It can be if the task is relatively clean.
But in speech synthesis, we usually adjust the range of f0 for each speaker.
In the case of ASR, it is difficult. So I'm not sure what is happened.

Comment (sw005320) [reply espnet2/tts/feats_extract/dio.py:26]: > > Do you think we can use it for ASR? If so, we'll test it.
> 
> It can be if the task is relatively clean.
> But in speech synthesis, we usually adjust the range of f0 for each speaker.
> In the case of ASR, it is difficult. So I'm not sure what will be happened.

I think the unadjusted range would be working as a sort of considering the speaker/environment characteristics, and it would be good.
I want to use a pythonic pitch implementation, so I'll try it.

File: egs2/jsut/tts1/conf/tuning/train_fastspeech.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
+# This configuration is for ESPnet2 to train FastSpeech.
+# It requires only a single GPU with 12 GB memory and it
+# takes ~2 days to finish the training on Titan V.
+
+# As default, we assume that the training is based on knowledge
```
Comment (sw005320) [inline egs2/jsut/tts1/conf/tuning/train_fastspeech.yaml:5]:
```suggestion
# As a default, we assume that the training is based on knowledge
```

File: espnet2/tts/feats_extract/dio.py [hunk 5]
```diff
@@ -23,7 +23,22 @@
...
+    """F0 estimation with dio + stonemask algortihm.
+
+    This is f0 extractor based on dio + stonmask algorithm introduced in `WORLD:
+    a vocoder-based high-quality speech synthesis system for real-time applications`_.
+
+    .. _`WORLD: a vocoder-based high-quality speech synthesis system for real-time
+        applications`: https://doi.org/10.1587/transinf.2015EDP7457
+
+    Note:
+        This module based on numpy implementaion. Therefore, the computational graph
```
Comment (sw005320) [inline espnet2/tts/feats_extract/dio.py:35]:
```suggestion
        This module is based on NumPy implementation. Therefore, the computational graph
```

File: egs2/jsut/tts1/conf/tuning/train_fastspeech.yaml (hunk 4 above)
Comment (kan-bayashi) [reply egs2/jsut/tts1/conf/tuning/train_fastspeech.yaml]: Sorry for many typo lol

File: espnet2/tts/feats_extract/dio.py (hunk 5 above)
Comment (kan-bayashi) [reply espnet2/tts/feats_extract/dio.py]: Sorry for many typo lol

File: espnet2/tts/fastspeech2.py [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+        duration_predictor_chans: int = 384,
+        duration_predictor_kernel_size: int = 3,
+        # energy predictor
+        energy_predictor_layers: int = 2,
+        energy_predictor_chans: int = 384,
+        energy_predictor_kernel_size: int = 3,
+        energy_predictor_dropout: float = 0.5,
+        energy_embed_kernel_size: int = 9,
+        energy_embed_dropout: float = 0.5,
+        stop_gradient_from_energy_predictor: bool = False,
```
Comment (r9y9) [inline espnet2/tts/fastspeech2.py:83]: I am not sure if this works well. Did you find it useful? 

I guess you are concerned about https://github.com/espnet/espnet/issues/2019#issuecomment-657189302. If the problem is the overfitting of variance adapter modules, I would turn loss weights and dropout rates.

File: espnet2/tts/fastspeech2.py [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+        # pitch predictor
+        pitch_predictor_layers: int = 2,
+        pitch_predictor_chans: int = 384,
+        pitch_predictor_kernel_size: int = 3,
+        pitch_predictor_dropout: float = 0.5,
+        pitch_embed_kernel_size: int = 9,
+        pitch_embed_dropout: float = 0.5,
+        stop_gradient_from_pitch_predictor: bool = False,
```
Comment (r9y9) [inline espnet2/tts/fastspeech2.py:91]: Same as above

**Comment by kan-bayashi** [comment]
> It seems that there is still room for improvement regarding prosody naturalness. Did you compare samples between 1) FastSpeech2 w/ ground-truth pitch/energy 2) FastSpeech2 w/ predicted pitch/energy? I am curious how close these samples are, and how close the first samples and ground-truth samples regarding prosody.

OK. I will add `use_teacher_forcing` option for FastSpeech.
Then, we can compare them.

**Comment by r9y9** [comment]
To clarify, what I meant by `there is room for improvement` is that there would be possibilities for future research to further enhance prosody.

The FastSpeech2 samples are definitely better than the ones of FastSpeech and that's very nice :+1:

File: espnet2/tts/feats_extract/energy.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        assert d.sum() == len(x)
+        d_cumsum = F.pad(d.cumsum(dim=0), (1, 0))
+        x_avg = [
+            x[start:end].masked_select(x[start:end].ne(0.0)).mean(dim=0)
+            if len(x[start:end].masked_select(x[start:end].ne(0.0))) != 0
+            else x.new_tensor(0.0)
+            for start, end in zip(d_cumsum[:-1], d_cumsum[1:])
+        ]
+        return torch.stack(x_avg)
+
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply espnet2/tts/feats_extract/energy.py]: I found that we do not need to check 0.0 > for energy.
I simplified this part.

File: espnet2/tts/variance_predictor.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            in_chans = idim if idx == 0 else n_chans
+            self.conv += [
+                torch.nn.Sequential(
+                    torch.nn.Conv1d(
+                        in_chans,
+                        n_chans,
+                        kernel_size,
+                        stride=1,
+                        padding=(kernel_size - 1) // 2,
+                    ),
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply espnet2/tts/variance_predictor.py]: In the case of `Conv1d -> ReLU -> LayerNorm`, bias may have slightly effect?

File: espnet2/tts/feats_extract/dio.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+        return f0
+
+    @staticmethod
+    def _average_by_duration(x: torch.Tensor, d: torch.Tensor) -> torch.Tensor:
+        assert d.sum() == len(x)
+        d_cumsum = F.pad(d.cumsum(dim=0), (1, 0))
+        x_avg = [
+            x[start:end].masked_select(x[start:end].ne(0.0)).mean(dim=0)
+            if len(x[start:end].masked_select(x[start:end].ne(0.0))) != 0
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply espnet2/tts/feats_extract/dio.py]: Changed.

File: espnet2/tts/fastspeech2.py (hunk 6 above)
Comment (kan-bayashi) [reply espnet2/tts/fastspeech2.py:83]: This option is derived from the advice from one of the authors.
Now I'm checking the effectiveness.

**Comment by kan-bayashi** [comment]
I added teacher forcing option for FastSpeech and FastSpeech2.
And I added samples with teacher forcing.
drive.google.com/drive/folders/18dZz5jP05Vh5bRRK8nT7HnmLXSeLsitm?usp=sharing

If we use groundtruth of pitch, pronunciation is almost perfect.

File: egs2/jsut/tts1/conf/tuning/train_fastspeech2.yaml [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    reduction_factor: 1               # reduction factor
+    init_type: xavier_uniform         # initialization type
+    init_enc_alpha: 1.0               # initial value of alpha of encoder scaled position encoding
+    init_dec_alpha: 1.0               # initial value of alpha of decoder scaled position encoding
+    transformer_enc_dropout_rate: 0.1            # dropout rate for transformer encoder layer
+    transformer_enc_positional_dropout_rate: 0.1 # dropout rate for transformer encoder positional encoding
+    transformer_enc_attn_dropout_rate: 0.1       # dropout rate for transformer encoder attention layer
+    transformer_dec_dropout_rate: 0.1            # dropout rate for transformer decoder layer
+    transformer_dec_positional_dropout_rate: 0.1 # dropout rate for transformer decoder positional encoding
+    transformer_dec_attn_dropout_rate: 0.1       # dropout rate for transformer decoder attention layer
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs2/jsut/tts1/conf/tuning/train_fastspeech2.yaml:45]: Thanks. I will try dropout rate = 0.2.

File: espnet2/tts/variance_predictor.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (r9y9) [reply espnet2/tts/variance_predictor.py]: Ah yes, right. I was worried about too much details.

File: espnet2/tts/fastspeech2.py (hunk 6 above)
Comment (r9y9) [reply espnet2/tts/fastspeech2.py:83]: Oops, I forgot about that. Thanks for the explanation.

File: egs2/jsut/tts1/conf/tuning/train_fastspeech2.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs2/jsut/tts1/conf/tuning/train_fastspeech2.yaml]: BTW, I also tried FastPitch setup (Lamb + Warmup + Weight decay) but it did not work well (loss keep the same value from the middle of the training). Maybe we need to tune LR.

**Comment by r9y9** [comment]
For pitch prediction, we may want to try discrete F0 (i.e., 0 as unvoiced regions) + quantization. Quantized F0 modeling has been successfully applied to SPSS and it's worth trying in my opinion. https://ieeexplore.ieee.org/document/8341752

**Comment by r9y9** [comment]
Looking at the FastSpeech 2 paper again, it seems that the authors use quantized one-hot F0 as input, but use scaler F0s as the output and optimize the pitch predictor by MSE loss. Umm, I don't understand why... I think we can use a cross entropy loss instead in the case of quantized F0 modeling.

**Comment by kan-bayashi** [comment]
>For pitch prediction, we may want to try discrete F0 (i.e., 0 in unvoiced regions) + quantization. Quantized F0 modeling has been successfully applied to SPSS and it's worth trying in my opinion. ieeexplore.ieee.org/document/8341752

Hmm. In the case of AR model, quantized f0 will be more useful, as we can see in vocoder?

**Comment by kan-bayashi** [comment]
> I don't understand why... I think we can use a cross entropy loss instead in the case of quantized F0 modeling.

About quantized f0 modeling, according to https://github.com/espnet/espnet/issues/2019#issuecomment-657173029, the accuracy seems not so high.

**Comment by dathudeptrai** [comment]
@kan-bayashi @r9y9 for FastSpeech2 paper, they use continuous f0 to train with MSE loss and use one-hot F0 as input to feed into the model. In the inference, they use real-value F0 outputs then transform it to one-hot f0 then fit it into the models. Use crossentropy is this case not make sense because the cross entropy (or accuracy) tell not thing about the variant of f0. It just like, groundtruth is category 1 but predict is category 200 or 2 has accuracy 0 %. But all we know that, 2 is better than 200.

**Comment by kan-bayashi** [comment]
@dathudeptrai Thank you for your explanation.
I agree that MSE and MAE can consider the distance (how far from the ground truth).

**Comment by kan-bayashi** [comment]
I think this PR is ready to merge.
I will merge this evening.

**Comment by sw005320** [comment]
Cool!
